common/base/include/claragenomics/utils/threadsafe_containers.hpp

```diff
@@ -0,0 +1,222 @@
+    {
+        size_t my_counter = counter_++;
+#ifdef __cpp_lib_optional
+        return my_counter < data_.size() ? cga_optional_t<T>(std::move(data_[my_counter])) : std::nullopt;
+#else
+        return my_counter < data_.size() ? cga_optional_t<T>(std::move(data_[my_counter])) : std::experimental::nullopt;
+#endif
+    }
+
+private:
+    /// data to provide, whenever an element is provided it is moved from this value
+    std::vector<T> data_;
+    /// number of elements provided so far
+    std::atomic<std::size_t> counter_;
+};
+
+/// ThreadsafeProducerConsumer - a threadsafe implementation of producer-consumer pattern with an option to signal that there will be no new elements
+///
+/// Producers add elements using add_new_element(), consumers consume elements in the order the were added using get_next_element().
+/// If there is no available element get_next_element() blocks and waits for one.
+///
+/// One producer can signal that there are not going to be any new elements using signal_pushed_last_element(). After this is done and all elements
+/// have been consumed get_next_element() returns empty optionals indicating that there is not going to be any new data and consumers can carry on
+template <typename T>
+class ThreadsafeProducerConsumer
+{
+public:
+    /// \brief default constructor
+    ThreadsafeProducerConsumer()
+        : data_()
+        , pushed_last_element_(false)
+        , mutex_()
+        , condition_variable_()
+    {
+    }
+
+    /// \brief deleted copy constructor
+    /// \param rhs
+    ThreadsafeProducerConsumer(const ThreadsafeProducerConsumer& rhs) = delete;
+
+    /// \brief deleted copy assignment operator
+    /// \param rhs
+    ThreadsafeProducerConsumer& operator==(const ThreadsafeProducerConsumer& rhs) = delete;
+
+    /// \brief deleted move constructor
+    /// \param rhs
+    ThreadsafeProducerConsumer(ThreadsafeProducerConsumer&& rhs) = delete; // could be implemeted if needed
+
+    /// \brief deleted move assignment operator
+    /// \param rhs
+    ThreadsafeProducerConsumer& operator==(ThreadsafeProducerConsumer&& rhs) = delete; // could be implemeted if needed
+
+    /// \brief destructor
+    ~ThreadsafeProducerConsumer() = default;
+
+    /// \brief adds an element to the queue
+    ///
+    /// \param element element to add
+    /// \throw std::logic_error if called after signal_pushed_last_element() has been called by any producer
+    void add_new_element(const T& element) // not using reference in order to support both rvalue and lvalue
+    {
+        // consider adding max number of elements and waiting for consumer to process an element before adding a new one
+        std::lock_guard<std::mutex> lg(mutex_);
+        if (pushed_last_element_)
+        {
+            throw std::logic_error("ThreadsafeProducerConsumer: pushed an element after signal_pushed_last_element() has been called");
+        }
+        data_.push_front(element);
+        condition_variable_.notify_all();
```

Comment:
I agree about `notify_one`.
I had a problem when `notify_x()` was outside of mutex, but I agree that should not happen. I'll look into it

Reply:
`test_threadsafe_producer_consumer_multiple_producers_multiple_consumers` takes 6s with `notify_all` and 10s with `notify_one`, but that's probably because in that case there are 100 producer threads constantly adding new value so I assume that `notify_all` wakes up multiple threads who then pick up values added by other producers, which in turn call `nofify_all` when there is no thread waiting which is probably faster than actually waking up threads one by one.
I'll anyway change it `notify_one` as it is the correct way to do it and this example is a bit extreme

Reply:
Done: switched to `notify_one()` and moved it outside of mutex

Reply:
Interesting, thanks for sharing the benchmark results. I never looked into performance of `notify_one` vs. `notify_all`. 100 threads doesn't sounds artifical, but not that extreme... Maybe there is a performance benefit. Nevertheless, I agree that `notify_one()` expresses the intent a bit better.

Reply:
With regards with `notify` inside and outside the lock

```
The notifying thread does not need to hold the lock on the same mutex as the one held by the waiting thread(s); in fact doing so is a pessimization, since the notified thread would immediately block again, waiting for the notifying thread to release the lock. However, some implementations (in particular many implementations of pthreads) recognize this situation and avoid this "hurry up and wait" scenario by transferring the waiting thread from the condition variable's queue directly to the queue of the mutex within the notify call, without waking it up. 
```

looks like some implementations make that optimization automatically. just read through it so thought i'd share